conn/gpio/gpiotest/gpiotest.go

```diff
@@ -128,5 +129,60 @@ func (p *PinPWM) PWM(duty gpio.Duty, period time.Duration) error {
 	return nil
 }
 
+// since returns time in µs since the test start.
+func since(start time.Time) string {
+	µs := (time.Since(start) + time.Microsecond/2) / time.Microsecond
+	ms := µs / 1000
+	µs %= 1000
+	return fmt.Sprintf("%3d.%03dms", ms, µs)
+}
+
+// LogPin logs when its state changes.
+type LogPin struct {
+	gpio.PinIO
+	start time.Time
+}
+
+// NewLogPin returns a new LogPin with current time.
+func NewLogPin(p gpio.PinIO) *LogPin {
+	return &LogPin{p, time.Now()}
+}
+
+// Real implements gpio.RealPin.
+func (p *LogPin) Real() gpio.PinIO {
+	return p.PinIO
+}
+
+// In implements gpio.PinIO.
+func (p *LogPin) In(pull gpio.Pull, edge gpio.Edge) error {
+	log.Printf("%s %s.In(%s, %s)\n", since(p.start), p, pull, edge)
+	return p.PinIO.In(pull, edge)
+}
+
+// Out mplements gpio.PinIO.
+func (p *LogPin) Out(l gpio.Level) error {
+	log.Printf("%s %s.Out(%s)\n", since(p.start), p, l)
+	return p.PinIO.Out(l)
+}
+
+// Read implements gpio.PinIO.
+func (p *LogPin) Read() gpio.Level {
+	l := p.PinIO.Read()
+	log.Printf("%s %s.Read() %s\n", since(p.start), p, l)
+	return l
+}
+
+// Pull implements gpio.PinIO.
+func (p *LogPin) Pull() gpio.Pull {
+	log.Printf("%s %s.Read()\n", since(p.start), p)
+	return p.PinIO.Pull()
+}
+
+// WaitForEdge implements gpio.PinIO.
+func (p *LogPin) WaitForEdge(timeout time.Duration) bool {
+	log.Printf("%s %s.WaitForEdge(%s)\n", since(p.start), p, timeout)
```

Comment:
I think it's more valuable to log after, e.g.

```
s := time.Now()
r := p.PinIO.WaitForEdge(timeout)
log.Printf("%s.WaitForEdge(%s) -> %t after %s", p, timeout, r, time.Since(s))
return r
```

That's probably the only case where I do see value to print out how long it waited.